sys/newlib_syscalls_default/syscalls.c

```diff
@@ -149,7 +144,23 @@ static const struct heap heaps[NUM_HEAPS] = {
  */
 void _init(void)
 {
-    /* nothing to do here */
+    /* Definition copied from newlib/libc/stdio/local.h */
+    extern void __sinit (struct _reent *);
+
+    /* When running multiple threads: Initialize reentrant structure before the
+     * scheduler starts. This normally happens upon the first stdio function
+     * called. However, if no boot message happens this can result in two
+     * concurrent "first calls" to stdio in data corruption, if no locking is
+     * used. Except for ESP (which is using its own syscalls.c anyway), this
+     * currently is the case in RIOT. */
+    if (MAXTHREADS > 1) {
+        /* Also, make an exception for riotboot, which does not use stdio
+         * at all. This would pull in stdio and increase .text size
+         * significantly there */
+        if (!IS_USED(MODULE_RIOTBOOT)) {
```

Comment:
I was under the impression that riotboot would already do

    DISBALE_MODULE += core_thread

but apparently not. That would save us some special casing in other places too.

Reply:
Would you open a PR to just do that? I could wait and rebase on top of that.

Reply:
`riotboot_dfu` uses a thread (and thread flags), so might to be so easy.
But how about

```suggestion
        if (!IS_USED(MODULE_STDIO_NULL)) {
```

Reply:
Sadly that would not work: `stdio_null` just throws away the output after it has gone the whole way through newlib. So there is no way for newlib to know whether the output is actually printed or just thrown away.

It would be possible to implement `stdio_null` to wrap all calls to `printf()` and friends to a mockup implementation. (And if we assume that callers of `printf()` don't care about the return value, those could be lean.) But as of know, we still need to init that with `stdio_null`.